### PR TITLE
detect/dataset: delay set operation after signature full match

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/5576

Describe changes:
- detect/dataset: delay set operation after signature full match

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2000

https://github.com/OISF/suricata/pull/11600 with patch working for multi buffers as well

Side note: the limitation described for flowvar in https://redmine.openinfosecfoundation.org/issues/7197 also applies here to dataset, and needs a bigger design...

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.

ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.

ERROR: QA failed on SURI_TLPR1_alerts_cmp.

ERROR: QA failed on IPS_AFP_drop_chk.
 | field | baseline | test | % |
|--|--|--|--| 
| | IPS_AFP_stats_chk | | |
| .ips.accepted | 367733210 | 342623210 | 93.17% |
| .ips.blocked | 747360 | 25857360 | 3459.83% |
| .ips.drop_reason.flow_drop | 680400 | 25736400 | 3782.54% |
| .ips.drop_reason.rules | 66960 | 120960 | 180.65% |
| .flow.end.state.established | 550799 | 604799 | 109.8% |
| .flow.end.state.closed | 1048672 | 994672 | 94.85% |
| .flow.end.tcp_state.established | 169560 | 223560 | 131.85% |
| .flow.end.tcp_state.closed | 1048672 | 994672 | 94.85% |
| .detect.alert | 185760 | 293760 | 158.14% |
| | TREX_GENERIC_stats_chk | | |
| .detect.alert | 86688 | 129368 | 149.23% | 

Pipeline 22065

**Comment by inashivb** [comment]
> Information:
> 
> ERROR: QA failed on SURI_TLPW2_single_alerts_cmp.
> 
> ERROR: QA failed on SURI_TLPW2_autofp_alerts_cmp.
> 
> ERROR: QA failed on SURI_TLPR1_alerts_cmp.
> 
> ERROR: QA failed on IPS_AFP_drop_chk.
> field 	baseline 	test 	%
> 	IPS_AFP_stats_chk 		
> .ips.accepted 	367733210 	342623210 	93.17%
> .ips.blocked 	747360 	25857360 	3459.83%
> .ips.drop_reason.flow_drop 	680400 	25736400 	3782.54%
> .ips.drop_reason.rules 	66960 	120960 	180.65%
> .flow.end.state.established 	550799 	604799 	109.8%
> .flow.end.state.closed 	1048672 	994672 	94.85%
> .flow.end.tcp_state.established 	169560 	223560 	131.85%
> .flow.end.tcp_state.closed 	1048672 	994672 	94.85%
> .detect.alert 	185760 	293760 	158.14%
> 	TREX_GENERIC_stats_chk 		
> .detect.alert 	86688 	129368 	149.23%
> 
> Pipeline 22065

Are these stat deviations expected, @catenacyber ?

File: src/detect-dataset.c [hunk 1]
```diff
@@ -41,8 +41,15 @@
 #include "util-path.h"
 #include "util-conf.h"
 
-int DetectDatasetMatch (ThreadVars *, DetectEngineThreadCtx *, Packet *,
-        const Signature *, const SigMatchCtx *);
+#define DMD_CAP_STEP 16
+typedef struct DetectDatasetMatchData_ {
+    uint32_t nb;
+    uint32_t cap;
```
Comment (inashivb) [inline src/detect-dataset.c:47]: Q: What do these indicate?

File: src/detect-dataset.c [hunk 2]
```diff
@@ -416,8 +489,25 @@ int DetectDatasetSetup (DetectEngineCtx *de_ctx, Signature *s, const char *rawst
     SCLogDebug("cmd %s, name %s",
         cmd_str, strlen(name) ? name : "(none)");
 
-    /* Okay so far so good, lets get this into a SigMatch
-     * and put it in the Signature. */
+    if (cmd == DETECT_DATASET_CMD_SET) {
+        // for set operation, we need one match, and one postmatch
```
Comment (inashivb) [inline src/detect-dataset.c:493]: Q: Why can't it be just a postmatch?

File: src/detect-dataset.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-dataset.c:47]: number of elements and capacity

File: src/detect-dataset.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-dataset.c:493]: postmatch does not have a direct access to the content of the sticky buffer(s) (+ transforms)...

Does that answer your question ?

Comment (inashivb) [reply src/detect-dataset.c:493]: > postmatch does not have a direct access to the content of the sticky buffer(s) (+ transforms)...

oh.

> Does that answer your question ?

yes. thank you very much! 🙇🏽‍♀️

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/11662

**Pull request closed without merging**